### DISCUSSION TITLE
accumulateTaggedlocalMatrix function in kernel

### DISCUSSION THREAD

**Discussion by xiaojbing** [idaholab/moose#20890] [Q&A General] [answered]
I want to inherit kernel.C and I want to know what is the function of accumulateTaggedlocalMatrix in kernel. The subclass must use this funcion or not? 

accumulateTaggedlocalMatrix in kernel.C: 
![image](https://user-images.githubusercontent.com/87591264/165512472-748e7bdb-7c25-4351-989d-d23d56e3cecd.png)

**Comment by loganharbour** [comment, marked as answer]
In general you should not override `Kernel::computeResidual()` or `Kernel::computeJacobian()`. You should override things like `Kernel::computeQp<>()`.

`accumulateTaggedLocalMatrix()` _must_ be called. It is what fills the local Jacobian into the global Jacobian based on what tags are set to be filled.

**Reply by xiaojbing** [reply]
> In general you should not override `Kernel::computeResidual()` or `Kernel::computeJacobian()`. You should override things like `Kernel::computeQp<>()`.
> 
> `accumulateTaggedLocalMatrix()` _must_ be called. It is what fills the local Jacobian into the global Jacobian based on what tags are set to be filled.

Thank you very much. I get it. But my code is not compute a variable's residuals or jacobian matrix，but a flux consisting of several variables. So there will be one or two more loops in these two functions: `computeResidual()` and `computeJacobian()`. Is a simple overloading possible for this case? If it is what you mean, is it necessary to implement the loop I want in `computeQp<>() `

**Reply by GiudGiud** [reply]
ok so you are making a custom kernel that is computing the contribution to the residual and jacobian for several variables?
Then you should override computeResidual, computeJacobian. But huge disclaimer that you are hacking MOOSE at this point. You will have to re-implement some of the stuff we made.

You should still follow the structure that we have: computeResidual calls computeQpResidual at each quadrature point and multiplies that contribution by the quadrature weight, volume, coord (for RZ) etc

accumulateTaggedLocalMatrix() transfers the contribution from the local thread to the global array. It does the work to avoid the race condition between threads etc.

**Reply by xiaojbing** [reply]
I followed the structure of MOOSE to build this custom kernel, but the results are not reasonable. If these two functions are overloaded, are there other files that need to be modified?

baseclass:
```
void
DGKernel::computeElemNeighJacobian(Moose::DGJacobianType type)
{
  const VariableTestValue & test_space =
      (type == Moose::ElementElement || type == Moose::ElementNeighbor) ? _test : _test_neighbor;
  const VariableTestValue & loc_phi =
      (type == Moose::ElementElement || type == Moose::NeighborElement) ? _phi : _phi_neighbor;

  if (type == Moose::ElementElement)
    prepareMatrixTag(_assembly, _var.number(), _var.number());
  else
    prepareMatrixTagNeighbor(_assembly, _var.number(), _var.number(), type);

  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
    for (_i = 0; _i < test_space.size(); _i++)
      for (_j = 0; _j < loc_phi.size(); _j++)
        _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeQpJacobian(type);

  accumulateTaggedLocalMatrix();

  if (_has_diag_save_in && (type == Moose::ElementElement || type == Moose::NeighborNeighbor))
  {
    unsigned int rows = _local_ke.m();
    DenseVector<Number> diag(rows);
    for (unsigned int i = 0; i < rows; i++)
      diag(i) = _local_ke(i, i);

    Threads::spin_mutex::scoped_lock lock(_jacoby_vars_mutex);
    for (const auto & var : _diag_save_in)
    {
      if (type == Moose::ElementElement)
        var->sys().solution().add_vector(diag, var->dofIndices());
      else
        var->sys().solution().add_vector(diag, var->dofIndicesNeighbor());
    }
  }
}
```

subclass:
```
void MultiDGKernel::computeElemNeighJacobian(Moose::DGJacobianType type)
{
  const VariableTestValue & test_space = (type == Moose::ElementElement || type == Moose::ElementNeighbor) ? _test : _test_neighbor;
  const VariableTestValue & loc_phi = (type == Moose::ElementElement || type == Moose::NeighborElement) ? _phi : _phi_neighbor;
  if (type == Moose::ElementElement)
    prepareMatrixTag(_assembly, _var.number(), _var.number());
  else
    prepareMatrixTagNeighbor(_assembly, _var.number(), _var.number(), type);

  for (_qp = 0; _qp<_qrule->n_points(); _qp++)
  {
 precalculateJacobian();
 for (unsigned int p = 0; p < _n_equation; ++p)
  for (unsigned int q = 0; q < _n_equation; ++q)
  {
   DenseMatrix<Number> & Kxx = type == Moose::ElementElement ?  
                                                            _assembly.jacobianBlock(p, q) : _assembly.jacobianBlockNeighbor(type, p, q);
   
   for (_i=0; _i<test_space.size(); _i++)
    for (_j=0; _j<loc_phi.size(); _j++)
     Kxx(_i, _j) += _JxW[_qp]*_coord[_qp]*computeQpJacobian(type, p, q);
  }
  }

  accumulateTaggedLocalMatrix();
  if (_has_diag_save_in && (type == Moose::ElementElement || type == Moose::NeighborNeighbor))
  {
    unsigned int rows = _local_ke.m();
    DenseVector<Number> diag(rows);
    for (unsigned int i = 0; i < rows; i++)
      diag(i) = _local_ke(i, i);
    Threads::spin_mutex::scoped_lock lock(_jacoby_vars_mutex);
    for (const auto & var : _diag_save_in)
    {
      if (type == Moose::ElementElement)
        var->sys().solution().add_vector(diag, var->dofIndices());
      else
        var->sys().solution().add_vector(diag, var->dofIndicesNeighbor());
    }
  }
}
```

**Reply by loganharbour** [reply]
I think we need more information on this statement: "I followed the structure of MOOSE to build this custom kernel, but the results are not reasonable."

Re-implementing all of the methods you are modifying above is not an easy task, and support for it is not something that we would typically offer on Discussions.
